### PR TITLE
Stat Panel: Support for custom prefixes and suffixes along with the standard formatting options

### PR DESCRIPTION
**Pull request by baldm0mma**
**What is this feature?**

This adds the ability for the user to add custom prefixes and suffixes to values in the Stat Panel, along with the standard support for unit formatting.

Also added a number of tests to our value formatting module.

**Why do we need this feature?**

It is a highly requested BI feature, and is table stakes for BI users.

**Who is this feature for?**

Everyone.

**Which issue(s) does this PR fix?**:

Addresses #63528

https://user-images.githubusercontent.com/46619047/222225631-b77e915d-1fc4-4a09-bb05-6a3d659d3cf6.mov

### CODE REVIEW

**Comment by ryantxu** [comment]
Interesting -- having this tied to percent feels odd since it is not actually calculating a percent.  Ideally the ↓ vs ↑ could be based on the change in the data, not a static config 🤔

**Comment by baldm0mma** [comment]
@ryantxu That's actually the hope, to build a better, more flexible answer to formatting values - I started a Design Doc for some possible solutions today - but this was an easy first step for a specific BI use case request.

**Comment by mdvictor** [comment]
I tend to agree with @ryantxu. I don't think the Unit option should have increase/decrease since they don't really represent units. But I do really like the idea of having this functionality on the stat panel though and it would definitely be cool to have it as a togglable option that shows the arrows based on some data change.

**Comment by ryantxu** [comment]
@baldm0mma -- here is a quick hack toying with where/how we could add a prefix based on the trend.   https://github.com/grafana/grafana/compare/stat-prefix

Obviously lots to figure out still... in particular what and where the options are!  Getting color picked based on the trend should also be considered

**Comment by baldm0mma** [comment]
Thanks @ryantxu ! This is also something my team talked about implementing, unfortunately the BI use cases we were shown don't involve comparing a previous value with a current value, instead the sql query generally returns a single value that the user wants to label themselves (usually in a Stat Panel), as either increasing or decreasing - or with a number of other prefixes (< / > / ~ / Δ / FQ_ / Qtr_ / FY____ etc) - that we currently have no option for anyway. I have a design doc [here](https://docs.google.com/document/d/1X7neLiaeHlh-nOo3RNgxGtp16eUE3OQ9BY7Paky2uXo/edit#) with a few options for better formatting. In fact, @codeincarnate and @mdvictor and I just talked this morning about implementing Proposal #1, which would basically create a new top-level BI category with a number of hard coded BI-realated prefix/suffix options. What do you think of that? EZ upgrade and get's the user what they need.

**Comment by torkelo** [comment]
I have been really wanting to add this to stat panel , a big thing it is missing . But I feel it should be an additional metric , not fully replace the value

**Comment by baldm0mma** [comment]
@torkelo do you mean something along the lines of keeping the current unit formatting the same, and in stat panel SPECIFICALLY having an option to either prepend or append one of the above-mentioned symbols to an already unit-formatted single value? Like, say we have the value `50`, we then use the unit dropdown to format it to a percentage unit (now `50%`), then have some way to then prepend (↑ / ↓ / < / > / + / ~ / Δ / etc) to the percentage (now ↑50%), OR any other chosen format (from `25 m/s2` -> `↑25 m/s2`)?

**Comment by torkelo** [comment]
Yes, like an additional metric below (including percentage change)

**Comment by baldm0mma** [comment]
Ah, I think I see what you mean; lemme build something in that direction.

**Comment by baldm0mma** [comment]
@ryantxu thanks for the idea to build this with text inputs; I think this gives the user a ton of flexible options, as well as is a lightweight solution that opens things up for more wide-ranging solutions in the future.

File: packages/grafana-data/src/field/fieldDisplay.ts [hunk 1]
```diff
@@ -64,6 +64,7 @@ export interface FieldDisplay {
   rowIndex?: number; // only filled in when the value is from a row (ie, not a reduction)
   getLinks?: () => LinkModel[];
   hasLinks: boolean;
+  fieldType?: FieldType;
```
Comment (ryantxu) [inline packages/grafana-data/src/field/fieldDisplay.ts:67]: is this just kicking the can... can we just add the field so we can use it later

Comment (ryantxu) [reply packages/grafana-data/src/field/fieldDisplay.ts:67]: maybe `source` or `sourceField`?

File: public/app/plugins/panel/stat/common.ts [hunk 2]
```diff
@@ -107,3 +111,55 @@ export function addOrientationOption<T extends SingleStatBaseOptions>(
...
+    }
+    return fieldValue;
+  });
+}
+
+export function addCustomUnitTextInputs(builder: FieldConfigEditorBuilder<FieldConfig>) {
+  const category = ['Custom units'];
+  return builder
+    .addTextInput({
+      path: 'customPrefix',
```
Comment (ryantxu) [inline public/app/plugins/panel/stat/common.ts:148]: We should add the new option to something like:
https://github.com/grafana/grafana/blob/main/public/app/plugins/panel/candlestick/panelcfg.cue#L27

something like:
```
PanelFieldConfig: {
  prefix?: string		
  suffix?: string					
} @cuetsy(kind="interface")
```

Given that it is already in "custom" i don't think we need/want it to be `custom.customPrefix`

File: packages/grafana-data/src/valueFormats/arithmeticFormatters.ts [hunk 3]
```diff
@@ -2,21 +2,21 @@ import { DecimalCount } from '../types/displayValue';
...
-export function toPercent(size: number, decimals: DecimalCount): FormattedValue {
+export function toPercent(size: number | null, decimals: DecimalCount): FormattedValue {
   if (size === null) {
     return { text: '' };
   }
   return { text: toFixed(size, decimals), suffix: '%' };
 }
 
-export function toPercentUnit(size: number, decimals: DecimalCount): FormattedValue {
+export function toPercentUnit(size: number | null, decimals: DecimalCount): FormattedValue {
```
Comment (ryantxu) [inline packages/grafana-data/src/valueFormats/arithmeticFormatters.ts:12]: maybe silly.. do you want to pull these changes into its own PR?  IIUC, this is just improving the types and tests -- and is not actually part of the StatsPanel changes

File: packages/grafana-data/src/field/fieldDisplay.ts (hunk 1 above)
Comment (baldm0mma) [reply packages/grafana-data/src/field/fieldDisplay.ts:67]: Good idea. I'll do that.

File: public/app/plugins/panel/stat/common.ts (hunk 2 above)
Comment (baldm0mma) [reply public/app/plugins/panel/stat/common.ts:148]: Yep, good catch. On it.

File: packages/grafana-data/src/valueFormats/arithmeticFormatters.ts (hunk 3 above)
Comment (baldm0mma) [reply packages/grafana-data/src/valueFormats/arithmeticFormatters.ts:12]: I was thinking the same thing, actually. Should have just done that. On it.

Comment (baldm0mma) [reply packages/grafana-data/src/valueFormats/arithmeticFormatters.ts]: removed.

File: packages/grafana-data/src/field/fieldDisplay.ts [hunk 4]
```diff
@@ -64,6 +64,8 @@ export interface FieldDisplay {
...
```
Comment (ryantxu) [inline packages/grafana-data/src/field/fieldDisplay.ts:67]: remove `fieldType` you can get it from `sourceField.type` :)

Comment (baldm0mma) [reply packages/grafana-data/src/field/fieldDisplay.ts:67]: LOL! Of course, sorry. On it.

File: public/app/plugins/panel/stat/common.ts [hunk 5]
```diff
@@ -107,3 +112,55 @@ export function addOrientationOption<T extends SingleStatBaseOptions>(
     defaultValue: VizOrientation.Auto,
   });
 }
+
+export function formatDisplayValuesWithCustomUnits(
+  fieldValues: FieldDisplay[],
+  customUnits: { customPrefix: string; customSuffix: string }
```
Comment (ryantxu) [inline public/app/plugins/panel/stat/common.ts:118]:
```suggestion
  config: PanelFieldConfig
```
can we just pass the options directly?

File: public/app/plugins/panel/stat/StatPanel.tsx [hunk 6]
```diff
@@ -82,6 +83,9 @@ export class StatPanel extends PureComponent<PanelProps<PanelOptions>> {
 
   getValues = (): FieldDisplay[] => {
     const { data, options, replaceVariables, fieldConfig, timeZone } = this.props;
+    // Test if there are custom units to prepend/append
+    const customPrefix = fieldConfig.defaults?.custom?.prefix ?? '';
+    const customSuffix = fieldConfig.defaults?.custom?.suffix ?? '';
```
Comment (ryantxu) [inline public/app/plugins/panel/stat/StatPanel.tsx:88]:
```suggestion
    const customConfig = fieldConfig.defaults?.custom ?? {};
```

File: public/app/plugins/panel/stat/StatPanel.tsx [hunk 7]
```diff
@@ -114,6 +118,16 @@ export class StatPanel extends PureComponent<PanelProps<PanelOptions>> {
       sparkline: options.graphMode !== BigValueGraphMode.None,
       timeZone,
     });
+
+    // If there are no custom prefixes or suffixes to format values, return fieldDisplayValues
+    if (!customPrefix && !customSuffix) {
```
Comment (ryantxu) [inline public/app/plugins/panel/stat/StatPanel.tsx:123]:
```suggestion
    if (customConfig.prefix || customConfig.suffix) {
       ... do the mapping....
```

File: packages/grafana-ui/src/components/Menu/Menu.tsx [hunk 8]
```diff
@@ -39,7 +39,16 @@ const MenuComp = React.forwardRef<HTMLDivElement, MenuProps>(
         aria-label={ariaLabel}
         onKeyDown={handleKeys}
       >
-        {header && <div className={styles.header}>{header}</div>}
+        {header && (
```
Comment (ryantxu) [inline packages/grafana-ui/src/components/Menu/Menu.tsx:42]: this seems unrelated

Comment (baldm0mma) [reply packages/grafana-ui/src/components/Menu/Menu.tsx:42]: I have no idea why this changed?

Comment (baldm0mma) [reply packages/grafana-ui/src/components/Menu/Menu.tsx:42]: Maybe I ran prettier?

File: public/app/plugins/panel/stat/common.ts (hunk 5 above)
Comment (baldm0mma) [reply public/app/plugins/panel/stat/common.ts:118]: Updated.

File: public/app/plugins/panel/stat/StatPanel.tsx (hunk 6 above)
Comment (baldm0mma) [reply public/app/plugins/panel/stat/StatPanel.tsx:88]: Updated.

File: packages/grafana-ui/src/components/Menu/Menu.tsx (hunk 8 above)
Comment (baldm0mma) [reply packages/grafana-ui/src/components/Menu/Menu.tsx]: somehow this got pulled into my branch; but i reverted it.

File: public/app/plugins/panel/stat/StatPanel.tsx [hunk 9]
```diff
@@ -114,6 +117,13 @@ export class StatPanel extends PureComponent<PanelProps<PanelOptions>> {
       sparkline: options.graphMode !== BigValueGraphMode.None,
       timeZone,
     });
+
+    // If there are custom prefixes or suffixes to format values, format the values
+    if (customConfig?.prefix || customConfig?.suffix) {
```
Comment (ryantxu) [inline public/app/plugins/panel/stat/StatPanel.tsx:122]: Pretty sure we can remove the null check since above we did `?? {}`
```suggestion
    if (customConfig.prefix || customConfig.suffix) {
```

Comment (baldm0mma) [reply public/app/plugins/panel/stat/StatPanel.tsx]: agreed

File: public/app/plugins/panel/stat/common.test.ts [hunk 10]
```diff
@@ -0,0 +1,113 @@
...
+      const unchangedFieldValues = formatDisplayValuesWithCustomUnits(mockFieldValues, {
+        // Empty custom values
+        prefix: '',
+        suffix: '',
+      });
+
+      // toEqual() recursively checks every field of an object or array for equality
+      expect(unchangedFieldValues).toEqual(mockFieldValues);
+    });
+    it('sucessfully prepends the `display.prefix` value with the custom prefix, and leave the suffix unchanged', () => {
```
Comment (nmarrs) [inline public/app/plugins/panel/stat/common.test.ts:53]: Super minot nit - I think having a line break between tests improves readability
```suggestion
    });

    it('sucessfully prepends the `display.prefix` value with the custom prefix, and leave the suffix unchanged', () => {
```

File: public/app/plugins/panel/stat/common.ts [hunk 11]
```diff
@@ -107,3 +112,55 @@ export function addOrientationOption<T extends SingleStatBaseOptions>(
...
+
+export function formatDisplayValuesWithCustomUnits(
+  fieldValues: FieldDisplay[],
+  config: PanelFieldConfig
+): FieldDisplay[] {
+  const { prefix, suffix } = config;
+
+  if (!prefix && !suffix) {
+    return fieldValues;
+  }
```
Comment (nmarrs) [inline public/app/plugins/panel/stat/common.ts:124]: Is this just a fail safe for usage in another panel if prefix / suffix aren't being checked? Seems a little redundant based on current usage / implementation 🤷‍♂️

File: public/app/plugins/panel/stat/common.test.ts [hunk 12]
```diff
@@ -0,0 +1,113 @@
+import { FieldType, FieldDisplay, ArrayVector } from '@grafana/data';
+
+import { formatDisplayValuesWithCustomUnits } from './common';
+
+describe('formatting functions', () => {
```
Comment (nmarrs) [inline public/app/plugins/panel/stat/common.test.ts:5]: Great job writing up test cases! 🥇

File: public/app/plugins/panel/stat/common.ts [hunk 13]
```diff
@@ -107,3 +112,55 @@ export function addOrientationOption<T extends SingleStatBaseOptions>(
...
+      return {
+        ...fieldValue,
+        display: {
+          ...display,
+          // Test also for nullishness, otherwise "undefined" will be concatenated into the string
+          prefix: `${prefix ?? ''}${display.prefix ?? ''}`,
+          suffix: `${display.suffix ?? ''}${suffix ?? ''}`,
+        },
+      };
+    }
```
Comment (nmarrs) [inline public/app/plugins/panel/stat/common.ts:140]: This might be a little cleaner if we split this out into two separate consts that got assigned as part of the returned object i.e.

```suggestion
    if (fieldType === FieldType.number) {
      const { display } = fieldValue;
      
      // Test also for nullishness, otherwise "undefined" will be concatenated into the string
      const customPrefix = `${prefix ?? ''}${display.prefix ?? ''}`
      const customSuffix = `${display.suffix ?? ''}${suffix ?? ''}`
      
      return {
        ...fieldValue,
        display: {
          ...display,
          prefix: customPrefix,
          suffix: customSuffix,
        },
      };
    }
```

**Review by nmarrs — APPROVED** [review]
This looks great! I am always a little hesitant with adding additional editor menus but barring including custom prefix / suffix as part of the standard options I think this gets us going in the right direction. Also given this only impacts stat panel for now I think we are good to have this implemented - if we want to apply this more broadly we will probably want to consider consolidation as potentially part of standard options

File: public/app/plugins/panel/stat/common.test.ts (hunk 10 above)
Comment (baldm0mma) [reply public/app/plugins/panel/stat/common.test.ts:53]: Agreed. On it.

File: public/app/plugins/panel/stat/common.test.ts (hunk 12 above)
Comment (baldm0mma) [reply public/app/plugins/panel/stat/common.test.ts:5]: Thanks!

File: public/app/plugins/panel/stat/common.ts (hunk 11 above)
Comment (baldm0mma) [reply public/app/plugins/panel/stat/common.ts:124]: I was thinking a fail safe. As you mentioned, I have a condition in `StatPanel.tsx` that only invokes this functions if there is at least one custom value. But if that condition isn't implemented alongside this function elsewhere, this will catch the fully-nullish conditions. Now that you mention it, it's probably best to remove the condition in StatPanel and just have the passthrough in this function. Thanks!

**Review by codeincarnate — APPROVED** [review]
Took a look and working well! I think the only downside at the moment is that suffixes are always displayed in a smaller font than the stat and the prefix. I'm pretty sure this was the pre-existing behavior but something we'll likely want to look at in the future as I doubt this assumption will hold across all suffixes that would be desirable to add.

Nice clean code! Some more TSDoc love would be great!

**Comment by ryantxu** [comment]
This is looking good -- I branched off this to see if we can tackle the long held request of making it data dependent -- with https://github.com/grafana/grafana/pull/64477/ it looks promising!

See:
https://github.com/grafana/grafana/compare/statsFormatting-alt

This branch includes a dev dashboard for this:



The config is pretty verbose, but not terrible:
image

File: public/app/plugins/panel/stat/StatPanel.tsx [hunk 14]
```diff
@@ -82,6 +83,8 @@ export class StatPanel extends PureComponent<PanelProps<PanelOptions>> {
 
   getValues = (): FieldDisplay[] => {
     const { data, options, replaceVariables, fieldConfig, timeZone } = this.props;
+    // Grab any custom units to prepend/append
+    const customConfig = fieldConfig.defaults?.custom ?? {};
```
Comment (ryantxu) [inline public/app/plugins/panel/stat/StatPanel.tsx:87]: this is not a safe way to check if field configs are applied... it only checks if the defaults are set, and does not check if the individual fields are set.

On top of that... the options ui will fill in default values for all of them anyway, so it will always be true 🤯

File: public/app/plugins/panel/stat/common.ts [hunk 15]
```diff
@@ -107,3 +112,60 @@ export function addOrientationOption<T extends SingleStatBaseOptions>(
     defaultValue: VizOrientation.Auto,
   });
 }
+
+export function formatDisplayValuesWithCustomUnits(
```
[the remark this answers is not in the document]
Comment (baldm0mma) [reply public/app/plugins/panel/stat/common.ts:116]: For sure! Updated.

File: public/app/plugins/panel/stat/common.ts [hunk 16]
```diff
@@ -107,3 +112,60 @@ export function addOrientationOption<T extends SingleStatBaseOptions>(
...
+          prefix: customPrefix,
+          suffix: customSuffix,
+        },
+      };
+    }
+    return fieldValue;
+  });
+}
+
+export function addCustomUnitTextInputs(builder: FieldConfigEditorBuilder<PanelFieldConfig>) {
```
[the remark this answers is not in the document]
Comment (baldm0mma) [reply public/app/plugins/panel/stat/common.ts:150]: Agreed. Updated.

File: public/app/plugins/panel/stat/StatPanel.tsx (hunk 14 above)
Comment (baldm0mma) [reply public/app/plugins/panel/stat/StatPanel.tsx]: Nice catch; I've updated the logic accordingly.

**Comment by baldm0mma** [comment]
Closing this pull request to work with @ryantxu on a data dependent override formatting solution. Picking up this work in #64764 .
